libs/src/services/solanaWeb3Manager/rewards.js

```diff
@@ -0,0 +1,767 @@
+ *   tokenAmount: BN
+ *   identityService: any
+ *   connection: Connection
+ * }} {
+ *   rewardManagerProgramId,
+ *   rewardManagerAccount,
+ *   attestations,
+ *   oracleAttestation,
+ *   challengeId,
+ *   specifier,
+ *   feePayer,
+ *   recipientEthAddress,
+ *   tokenAmount,
+ *   identityService,
+ *   connection
+ * }
+ */
+export async function submitAttestations({
+  rewardManagerProgramId,
+  rewardManagerAccount,
+  attestations,
+  oracleAttestation,
+  challengeId,
+  specifier,
+  feePayer,
+  recipientEthAddress,
+  tokenAmount,
+  identityService,
+  connection,
+}) {
+  // Construct combined transfer ID
+  const transferId = constructTransferId(challengeId, specifier)
+
+  // Derive the message account we'll use to store the attestations
+  const [
+    rewardManagerAuthority,
+    derivedMessageAccount,
+  ] = await deriveMessageAccount(
+    transferId,
+    rewardManagerProgramId,
+    rewardManagerAccount
+  )
+
+  // Add instructions from DN attestations - each attestation
+  // needs a pairing of SECP recovery instruction and submit
+  // attestation instruction.
+  let instructions = await Promise.all(
+    attestations.reduce((instructions, meta, i) => {
+      const verifyInstruction = generateSubmitAttestationInstruction({
+        attestationMeta: meta,
+        derivedMessageAccount,
+        rewardManagerAccount,
+        rewardManagerProgramId,
+        rewardManagerAuthority,
+        transferId,
+        feePayer,
+      })
+      const secpInstruction = Promise.resolve(
+        generateSecpInstruction({
+          attestationMeta: meta,
+          isOracle: false,
+          recipientEthAddress,
+          oracleAddress: oracleAttestation.ethAddress,
+          tokenAmount,
+          transferId,
+          instructionIndex: 2 * i,
+        })
+      )
+      return [...instructions, secpInstruction, verifyInstruction]
+    }, [])
+  )
+
+  // Add instructions from oracle attestation
+  const oracleSecp = await generateSecpInstruction({
+    attestationMeta: oracleAttestation,
+    recipientEthAddress,
+    instructionIndex: instructions.length,
+    transferId,
+    isOracle: true,
+    tokenAmount,
+    oracleAddress: oracleAttestation.ethAddress,
+  })
+  const oracleTransfer = await generateSubmitAttestationInstruction({
+    attestationMeta: oracleAttestation,
+    derivedMessageAccount,
+    rewardManagerAccount,
+    rewardManagerProgramId,
+    rewardManagerAuthority,
+    transferId,
+    feePayer,
+  })
+  instructions = [...instructions, oracleSecp, oracleTransfer]
+
+  // Prep transaction to be relayed and send it
+  const relayable = instructions.map(prepareInstructionForRelay)
+  const { blockhash: recentBlockhash } = await connection.getRecentBlockhash()
+  const transactionData = {
+    recentBlockhash,
+    instructions: relayable,
+  }
+
+  try {
+    const response = await identityService.solanaRelay(transactionData)
+    return response
+  } catch (e) {
+    console.error(e.message)
```

Comment:
worth rethrowing if we're not going to return anything?

Reply:
yeah - need to do a whole pass on errors where we make identity return the actual error and map to known types in libs. going to leave as is for now until i do this work